File: web/js/animations.js
// Port of src/animations.cpp, kept deliberately line-for-line.
//
// Both sides are pure functions of elapsed time, so the preview she sees while
// choosing an animation is the same frame the matrix will draw at that
// millisecond. If you change one file, change the other.

export const COLUMNS = 32;

// Sprites as column bytes, bit 0 = top row. Identical to animations.cpp.

// .##...##. / ######### / ######### / ######### / .#######. / ..#####.. / ...###... / ....#....
const HEART_BIG = [0x0e, 0x1f, 0x3f, 0x7e, 0xfe, 0x7e, 0x3f, 0x1f, 0x0e];

// ....... / .##.##. / ####### / ####### / .#####. / ..###.. / ...#... / .......
const HEART_MID = [0x0c, 0x1e, 0x3e, 0x7c, 0x3e, 0x1e, 0x0c];

// ..... / ..... / .#.#. / ##### / ##### / .###. / ..#.. / .....
const HEART_SMALL = [0x18, 0x3c, 0x78, 0x3c, 0x18];

// ...... / ...... / ....#. / .....# / ###### / .....# / ....#. / ......
const ARROW = [0x10, 0x10, 0x10, 0x10, 0x54, 0x38];

/** Ids as sent over BLE, with the Polish labels shown in the gallery. */
export const ANIMATIONS = [
  { id: 'heartbeat', label: 'Bicie serca', interval: 40 },
  { id: 'hearts', label: 'Serduszka', interval: 55 },
  { id: 'sparkle', label: 'Gwiazdki', interval: 110 },
  { id: 'arrow', label: 'Strzała', interval: 40 },
];

export function frameInterval(id) {
  return ANIMATIONS.find((a) => a.id === id)?.interval ?? 40;
}

function clear(frame) {
  frame.fill(0);
}

/** Draws a sprite with its left edge at column x, clipped at both ends. */
function blit(frame, sprite, x) {
  for (let i = 0; i < sprite.length; i++) {
    const col = x + i;
    if (col >= 0 && col < COLUMNS) {
      frame[col] |= sprite[i];
    }
  }
}

function blitCentred(frame, sprite) {
  blit(frame, sprite, Math.floor((COLUMNS - sprite.length) / 2));
}

/** Matches hash32() in animations.cpp, so the sparkle pattern is identical. */
function hash32(x) {
  x = Math.imul(x ^ (x >>> 16), 0x7feb352d);
  x = Math.imul(x ^ (x >>> 15), 0x846ca68b);
  return (x ^ (x >>> 16)) >>> 0;
}

// Real hearts go lub-DUB ... pause. Evenly spaced pulses just read as blinking.
function renderHeartbeat(frame, elapsedMs) {
  const t = elapsedMs % 1300;
  clear(frame);
  if (t < 130) blitCentred(frame, HEART_BIG);
  else if (t < 260) blitCentred(frame, HEART_MID);
  else if (t < 400) blitCentred(frame, HEART_BIG);
  else if (t < 520) blitCentred(frame, HEART_MID);
  else blitCentred(frame, HEART_SMALL);
}

function renderHearts(frame, elapsedMs) {
  const spacing = 13;
  const span = spacing * 3;
  clear(frame);
  const drift = Math.floor(elapsedMs / 55) % span;
  for (let i = -1; i < 4; i++) {
    blit(frame, HEART_MID, COLUMNS - drift + i * spacing);
  }
}

function renderSparkle(frame, elapsedMs) {
  const slot = Math.floor(elapsedMs / 110);
  clear(frame);
  for (let col = 0; col < COLUMNS; col++) {
    for (let row = 0; row < 8; row++) {
      // Each star lives two slots, so they fade rather than strobe.
      const a = hash32((col * 911 + row * 3571 + slot * 7919) >>> 0);
      const b = hash32((col * 911 + row * 3571 + (slot - 1) * 7919) >>> 0);
      if ((a & 0xff) < 14 || (b & 0xff) < 8) {
        frame[col] |= 1 << row;
      }
    }
  }
}

function renderArrow(frame, elapsedMs) {
  const period = 2600;
  const flight = 1400;
  const start = -ARROW.length;
  const end = COLUMNS;

  const t = elapsedMs % period;
  clear(frame);

  const struck = t > (flight * 50) / 100 && t < (flight * 75) / 100;
  if (struck) {
    blitCentred(frame, HEART_BIG);
  } else {
    blit(frame, HEART_MID, Math.floor((COLUMNS - HEART_MID.length) / 2));
  }

  if (t < flight) {
    blit(frame, ARROW, start + Math.floor((t * (end - start)) / flight));
  }
}

/**
 * Renders one frame into a Uint8Array(32). Overwrites every column.
 * Unknown ids fall back to heartbeat, as the firmware does.
 *
 * Named renderAnimation rather than render so it stays unambiguous alongside
 * preview.js's own drawing functions when the modules are flattened into a
 * single file (see tools/build_single_file.py).
 */
export function renderAnimation(id, frame, elapsedMs) {
  switch (id) {
    case 'hearts': return renderHearts(frame, elapsedMs);
    case 'sparkle': return renderSparkle(frame, elapsedMs);
    case 'arrow': return renderArrow(frame, elapsedMs);
    default: return renderHeartbeat(frame, elapsedMs);
  }
}

File: src/animations.cpp
#include "animations.h"

#include <string.h>

namespace anim {
namespace {

// --- sprites ----------------------------------------------------------------
//
// Column bytes, bit 0 = top row. The ASCII art above each one is the source of
// truth for what it should look like; the numbers are derived from it.

// .##...##.
// #########
// #########
// #########
// .#######.
// ..#####..
// ...###...
// ....#....
constexpr uint8_t kHeartBig[] = {0x0E, 0x1F, 0x3F, 0x7E, 0xFE, 0x7E, 0x3F, 0x1F, 0x0E};

// .......
// .##.##.
// #######
// #######
// .#####.
// ..###..
// ...#...
// .......
constexpr uint8_t kHeartMid[] = {0x0C, 0x1E, 0x3E, 0x7C, 0x3E, 0x1E, 0x0C};

// .....
// .....
// .#.#.
// #####
// #####
// .###.
// ..#..
// .....
constexpr uint8_t kHeartSmall[] = {0x18, 0x3C, 0x78, 0x3C, 0x18};

// ......
// ......
// ....#.
// .....#
// ######
// .....#
// ....#.
// ......
constexpr uint8_t kArrow[] = {0x10, 0x10, 0x10, 0x10, 0x54, 0x38};

constexpr size_t kHeartBigW = sizeof(kHeartBig);
constexpr size_t kHeartMidW = sizeof(kHeartMid);
constexpr size_t kHeartSmallW = sizeof(kHeartSmall);
constexpr size_t kArrowW = sizeof(kArrow);

// --- helpers ----------------------------------------------------------------

void clear(uint8_t frame[kColumns]) { memset(frame, 0, kColumns); }

// Draws a sprite with its left edge at column `x`, OR-ed into the frame and
// clipped at both edges. `x` is signed so sprites can walk in from off-screen.
void blit(uint8_t frame[kColumns], const uint8_t *sprite, size_t width, int32_t x) {
  for (size_t i = 0; i < width; i++) {
    const int32_t col = x + static_cast<int32_t>(i);
    if (col >= 0 && col < static_cast<int32_t>(kColumns)) {
      frame[col] |= sprite[i];
    }
  }
}

void blitCentred(uint8_t frame[kColumns], const uint8_t *sprite, size_t width) {
  blit(frame, sprite, width, static_cast<int32_t>((kColumns - width) / 2));
}

// A cheap integer hash, so the sparkle pattern is random-looking but is still
// a pure function of position and time.
uint32_t hash32(uint32_t x) {
  x ^= x >> 16;
  x *= 0x7FEB352DU;
  x ^= x >> 15;
  x *= 0x846CA68BU;
  x ^= x >> 16;
  return x;
}

// --- the animations ---------------------------------------------------------

// Real hearts go lub-DUB ... pause. Two beats close together, then rest, which
// reads as a heartbeat where an evenly-spaced pulse just reads as blinking.
void renderHeartbeat(uint8_t frame[kColumns], uint32_t elapsedMs) {
  constexpr uint32_t kPeriod = 1300;
  const uint32_t t = elapsedMs % kPeriod;

  clear(frame);
  if (t < 130) {
    blitCentred(frame, kHeartBig, kHeartBigW);  // lub
  } else if (t < 260) {
    blitCentred(frame, kHeartMid, kHeartMidW);
  } else if (t < 400) {
    blitCentred(frame, kHeartBig, kHeartBigW);  // DUB
  } else if (t < 520) {
    blitCentred(frame, kHeartMid, kHeartMidW);
  } else {
    blitCentred(frame, kHeartSmall, kHeartSmallW);  // rest
  }
}

// Three hearts drifting right to left at even spacing, wrapping seamlessly.
void renderHearts(uint8_t frame[kColumns], uint32_t elapsedMs) {
  constexpr int32_t kSpacing = 13;
  constexpr int32_t kSpan = kSpacing * 3;  // distance before the pattern repeats
  constexpr uint32_t kMsPerColumn = 55;

  clear(frame);
  const int32_t drift = static_cast<int32_t>((elapsedMs / kMsPerColumn) % kSpan);

  // Start one spacing off the right edge so hearts enter rather than pop in.
  for (int32_t i = -1; i < 4; i++) {
    blit(frame, kHeartMid, kHeartMidW, kColumns - drift + i * kSpacing);
  }
}

void renderSparkle(uint8_t frame[kColumns], uint32_t elapsedMs) {
  constexpr uint32_t kSlotMs = 110;
  const uint32_t slot = elapsedMs / kSlotMs;

  clear(frame);
  for (uint32_t col = 0; col < kColumns; col++) {
    for (uint32_t row = 0; row < 8; row++) {
      // Each star lives for two slots, so they fade in and out rather than
      // strobing every frame.
      const uint32_t a = hash32(col * 911u + row * 3571u + slot * 7919u);
      const uint32_t b = hash32(col * 911u + row * 3571u + (slot - 1) * 7919u);
      if ((a & 0xFF) < 14 || (b & 0xFF) < 8) {
        frame[col] |= static_cast<uint8_t>(1u << row);
      }
    }
  }
}

// An arrow crosses the display and buries itself in a heart, which flinches.
void renderArrow(uint8_t frame[kColumns], uint32_t elapsedMs) {
  constexpr uint32_t kPeriod = 2600;
  constexpr int32_t kStart = -static_cast<int32_t>(kArrowW);
  constexpr int32_t kEnd = kColumns;
  constexpr uint32_t kFlightMs = 1400;

  const uint32_t t = elapsedMs % kPeriod;
  clear(frame);

  // The heart is struck once the arrow's tip reaches the middle.
  const int32_t heartLeft = static_cast<int32_t>((kColumns - kHeartMidW) / 2);
  const bool struck = t > (kFlightMs * 50 / 100) && t < (kFlightMs * 75 / 100);

  if (struck) {
    blitCentred(frame, kHeartBig, kHeartBigW);
  } else {
    blit(frame, kHeartMid, kHeartMidW, heartLeft);
  }

  if (t < kFlightMs) {
    const int32_t travel = kEnd - kStart;
    const int32_t x = kStart + static_cast<int32_t>(t * travel / kFlightMs);
    blit(frame, kArrow, kArrowW, x);
  }
}

}  // namespace

// --- public -----------------------------------------------------------------

const char *toString(Id id) {
  switch (id) {
    case Id::Heartbeat: return "heartbeat";
    case Id::Hearts: return "hearts";
    case Id::Sparkle: return "sparkle";
    case Id::Arrow: return "arrow";
    default: return "heartbeat";
  }
}

Id fromString(const char *id) {
  if (id == nullptr) {
    return Id::Heartbeat;
  }
  for (uint8_t i = 0; i < static_cast<uint8_t>(Id::Count); i++) {
    const Id candidate = static_cast<Id>(i);
    if (strcmp(id, toString(candidate)) == 0) {
      return candidate;
    }
  }
  return Id::Heartbeat;
}

void render(Id id, uint8_t frame[kColumns], uint32_t elapsedMs) {
  switch (id) {
    case Id::Hearts: renderHearts(frame, elapsedMs); break;
    case Id::Sparkle: renderSparkle(frame, elapsedMs); break;
    case Id::Arrow: renderArrow(frame, elapsedMs); break;
    case Id::Heartbeat:
    default: renderHeartbeat(frame, elapsedMs); break;
  }
}

uint16_t frameInterval(Id id) {
  switch (id) {
    case Id::Hearts: return 55;
    case Id::Sparkle: return 110;
    case Id::Arrow: return 40;
    case Id::Heartbeat:
    default: return 40;
  }
}

}  // namespace anim
